File: src/condition/ResourceMatch.php
<?php


namespace Ething\Condition;

use \Ething\ResourceQueryParser;
	
class ResourceMatch extends AbstractResourceCondition {
	
	
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 
				'resource' => null,
				'expression' => null
			),
			$attributes
		);
		
		foreach($attributes as $key => &$value){
			
			switch($key){
				
				case 'resource':
					
					try {
						static::validateResourceAttr($context['ething'], $value);
					} catch(\Exception $e){
						throw new \Exception("[{$key}] ".$e->getMessage(),$e->getCode(),$e);
					}
					break;
				
				case 'expression':
					
					if(!is_string($value) || empty(trim($value)))
						throw new \Exception("field '{$key}' must be a non empty string.");
					
					// parse the query string
					if(!ResourceQueryParser::check($context['ething'], $value, $error))
						throw new \Exception('invalid expression: '.$error);
					
					break;
				
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true; 
	}
	
	protected function call(\Ething\Event\Signal $signal){
		$resources = $this->getResources($signal);
		
		foreach($resources as $r){
			return $r->match($this->expression);
		}
	}
	
	
}

File: ething/rule/condition/ResourceMatch.py
from .ResourceCondition import ResourceCondition
from ething.ResourceQueryParser import ResourceQueryParser


class ResourceMatch(ResourceCondition):
	
	@staticmethod
	def validate (attributes, context):
		
		attributes.setdefault('expression', None)
		
		other = {}
		
		for key in attributes:
			
				
			if key == 'expression':
				
				expr = attributes[key]
				
				if not( isinstance(expr, basestring) and len(expr)>0 ):
					raise Exception("must be a non empty string.")
				
				ok, message = ResourceQueryParser.check(value)
				
				if not ok:
					raise Exception("invalid expression: %s" % message)
				
				
			else:
				other[key] = attributes[key]
		
		return ResourceCondition.validate(other, context)
	
	
	
	def call(self, signal):
		
		resources = self.getResourcesFromSignal(signal)
		
		for r in resources:
			if not r.match(self['expression']):
				return False
		
		return True
	